articles/external_qos_file.md

```diff
@@ -0,0 +1,402 @@
+
+ROS 2 can use a simpler approach, leveraging node name uniqueness <sup id="back_node_name_uniqueness">[1](#to_node_name_uniqueness)</sup>.
+
+<b id="to_node_name_uniqueness">1</b> Node name uniqueness is not being enforced up to Foxy, though it is supposed. [↩](#back_node_name_uniqueness)
+
+### Nodes
+
+In an acceptable `QoS` file format, it should be to distinguish `QoS` settings targeted to different nodes.
+Thus, a top level tag to identify the node should be available.
+Example alternatives:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        ...
+    </node>
+<qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        ...
+```
+
+### Endpoint type
+
+It should be possible to distinguish between the different endpoint types in a qos profile file.
+For example:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        <publisher topic_name="asd">  <!--Relative name-->
+            ...
+        </publisher>
+        <subscription topic_name="/another_ns/asd">  <!--Absolute name-->
+            ...
+        </subscription>
+    </node>
+<qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        publisher:
+            topic_name: asd
+            qos:
+                ...
+        subscription:
+            topic_name: /another_ns/asd
+            qos:
+                ...
+```
+
+Besides `publisher` and `subscription` tags, `client` and `service` should be allowed.
+
+### QoS profiles ids (optional)
+
+In the case that a node wants to create two `publishers` on the same topic with different QoS, the above format wouldn't allow to identify uniquely the `publisher`. The same applies to other kinds of endpoints.
+
+To solve that issue, the format could allow an using a profile id, instead of the topic name:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        <publisher topic_name="asd" profile_id="1">
+            ...
+        </publisher>
+        <subscription topic_name="asd" profile_id="2">
+            ...
+        </subscription>
+    </node>
+<qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        publisher:
+            topic_name: asd
+            profile_id: 1
+            qos:
+                ...
+        subscription:
+            topic_name: asd
+            profile_id: 2
+            qos:
+                ...
+```
+
+`rcl` API would need to be extended to support profiles id.
+This mechanism should be avoided, except in the few use cases where there are collisions.
+
+### Default profiles for endpoints in different nodes
+
+It is common to have several nodes creating the same endpoint, and the same qos profile is desired in all of them.
+A default tag can be added to solve this:
```

Comment:
It's not clear when this default is applied, probably because it's not (yet) clear when the profiles in this file are available. Are they loaded by a node, or the launch system, or is the entire system restricted to one file?

Reply:
> probably because it's not (yet) clear when the profiles in this file are available

My first idea was to fully specify the QoS profile iniline after the entity tag:

```xml
<qos_profiles>
    <node name="my_node" namespace="/my_ns/nested_ns">
        <publisher topic_name="asd">
            <qos base_profile="default">  <!--Defaults to rmw_qos_profile_t default-->
              <reliability>reliable</reliability>
              <history_depth>100</history_depth>
            </qos>
        </publisher>
    </node>
</qos_profiles>
```

But I think that having a way to define a profile in the file and reusing it is a better idea (defining the profile inline should still be valid).

> It's not clear when this default is applied

If there's no rule specific for the node looking up the file, it will use the one in `default`.
I also wanted to have a way of only overriding some of the QoS settings, and use as a base the profile in default e.g.:

```xml
<qos_profiles>
    <default>
        <publisher topic_name="asd">
            <qos base_profile="default">
              <reliability>best_effort</reliability>
              <history_depth>100</history_depth>
            </qos>
        </publisher>
    </default>

    <node name="my_node" namespace="/my_ns/nested_ns">
        <publisher topic_name="asd">
            <qos>
              <!--This is already best effort, as the profile in <default> says so-->
              <history_depth>10</history_depth>
            </qos>
        </publisher>
    </node>
</qos_profiles>
```
though that might be confusing.

---

> Are they loaded by a node, or the launch system, or is the entire system restricted to one file?

The launch system doesn't load anything.
The proposal includes `rcl` API for loading the file.
There would be one global loaded file in init_options, and each node can optionally load a different file.

See `Usage in init options and node options` section.
(it's the same idea that having both "global arguments" and "local arguments").

The structure that represents the "loaded file" could also be filled up by other means, so the side loading system won't be restricted to a file format.

Reply:
I think this is clearer now after https://github.com/ros2/design/pull/296/commits/a0e635f3500e457bf56a36862651fe3bcf50b0a7.